### PR TITLE
Monitoring

### PR DESCRIPTION
**Pull request by akshayknarayan**
Basic monitoring functionality.

### CODE REVIEW

File: minion/monitor/monitor.go [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+		log.WithError(err).Warn("container not found")
+		return false
+	}
+	out, err := dk.ExecVerbose(dCont.Name[1:], "ping", "-q", "-c3", ip)
+	if err != nil {
+		log.WithError(err).Warn(string(out))
+		return false
+	}
+
+	lines := strings.Split(string(out), "\n")
```
Comment (kklin) [inline minion/monitor/monitor.go]: Could we just use the exit code of ping?

File: minion/monitor/monitor.go [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+
+type connection struct {
+	from db.Container
+	to   db.Container
+}
+
+// Monitor checks via ping that each connection outlined in the stitch is live.
+func Monitor(conn db.Conn, dckr docker.Client) {
+	tick := time.Tick(15 * time.Second)
+	for {
```
Comment (kklin) [inline minion/monitor/monitor.go]: Would be cleaner if you `range`d over the tick

File: minion/docker/docker.go [hunk 3]
```diff
@@ -153,26 +154,32 @@ func (dk Client) ExecVerbose(name string, cmd ...string) ([]byte, error) {
...
 	for scanner.Scan() {
 		outBuff.WriteString(scanner.Text() + "\n")
 	}
 	if err := scanner.Err(); err != nil {
-		return nil, err
+		return nil, -1, err
 	}
 
-	return outBuff.Bytes(), nil
+	return outBuff.Bytes(), exitCode, nil
```
Comment (secant) [inline minion/docker/docker.go]: (Nitpicky) You could return `execInspect.ExitCode` and not declare `exitCode`

File: minion/monitor/monitor.go [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+)
+
+type connection struct {
+	from db.Container
+	to   db.Container
+}
+
+// Monitor checks via ping that each connection outlined in the stitch is live.
+func Monitor(conn db.Conn, dckr docker.Client) {
+	for tick := time.Tick(15 * time.Second); true; <-tick {
```
Comment (secant) [inline minion/monitor/monitor.go]: Can be written as `for tick := range time.Tick(15 * time.Second)`

File: minion/monitor/monitor.go [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+	if err != nil {
+		log.WithError(err).Warn("[monitor] Invalid spec.")
+		return
+	}
+
+	var conns []connection
+	for _, e := range edges {
+		fr, err := nodeToContainer(e.From, containers)
+		if err != nil {
+			log.WithError(err).Warn("[monitor] Could not find container.")
```
Comment (secant) [inline minion/monitor/monitor.go]: Maybe specify which container could not be found?

File: minion/monitor/monitor.go [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+		log.Warn(fmt.Sprintf("[monitor] %s -> %s unreachable!",
+			fail.from.IP, fail.to.IP))
+	}
+}
+
+func getEdges(conn db.Conn) ([]stitch.Edge, error) {
+	var minion db.Minion
+	tick := time.Tick(30 * time.Second)
+	for {
+		<-tick
```
Comment (secant) [inline minion/monitor/monitor.go]: `time.Sleep(30 * time.Second)` instead?

**Comment by ejj** [comment]
Let's shelve this for now.  We didn't use it in the paper and I don't see an immediate application for our early target users.  Perhaps we should put it on a branch?  Not sure.

**Pull request closed without merging**